src/player.cpp

```diff
@@ -11158,162 +11158,184 @@ void player::gunmod_add( item &gun, item &mod )
 
 hint_rating player::rate_action_read( const item &it ) const
 {
-    if (!it.is_book()) {
+    if( !it.is_book() ) {
         return HINT_CANT;
     }
 
-    //Check for NPCs to read for you, negates Illiterate and Far Sighted
-    //The NPC gets a slight boost to int requirement since they wouldn't need to
-    //understand what they are reading necessarily
-    int assistants = 0;
-    for( auto &elem : g->active_npc ) {
-        if (rl_dist( elem->pos(), g->u.pos() ) < PICKUP_RANGE && elem->is_friend()) {
-            ///\EFFECT_INT_NPC allows NPCs to read harder books for you
-            if ((elem->int_cur+1) >= it.type->book->intel) {
-                if ( elem->in_sleep_state() ) {
-                    //Sleeping NPCs would need to be awoken first...
-                    //Tell the player about it.
-                    add_msg(m_info, _("%s could read this to you, if they were awake..."), elem->name.c_str());
-                } else {
-                    assistants++;
-                }
-            }
-       }
-    }
-
-    if (g && g->m.ambient_light_at(pos()) < 8 && LL_LIT > g->m.light_at(pos())) {
-        return HINT_IFFY;
-    } else if( !has_morale_to_read() && it.type->book->fun <= 0 ) {
-        return HINT_IFFY; //won't read non-fun books when sad
-    } else if (it.type->book->intel > 0 && has_trait("ILLITERATE") && assistants == 0) {
-        return HINT_IFFY;
-    } else if (has_trait("HYPEROPIC") && !is_wearing("glasses_reading") &&
-               !is_wearing("glasses_bifocal") && !has_effect( effect_contacts ) && assistants == 0) {
-        return HINT_IFFY;
-    }
-
-    return HINT_GOOD;
+    std::vector<std::string> dummy;
+    const player *dummy_pl;
+    return time_to_read( it, dummy_pl, dummy ) == INT_MAX ? HINT_IFFY : HINT_GOOD;
 }
 
-bool player::read(int inventory_position)
+int player::time_to_read( const item &book, player const *&reader,
+                          std::vector<std::string> &reasons ) const
```

Comment:
This pointer return thing is pretty ugly.
It can be cleanly split into 2 functions: `time_to_read` and something like `get_book_reader`. First one would take arguments similar to this version, except the `reader` would be a reference, the second one would take just the book and `reasons` vector.